upgrade to zig 0.14.0

chrisirhc opened the pull request:
Disclaimer: No idea what I'm doing.

chrisirhc committed:
upgrade to zig 0.13.0

chrisirhc committed:
fix windows

chrisirhc committed:
attempt 0.14 nightly

chrisirhc commented:
Hit compilation errors which indicate that threads weren't turned on:

```
In file included from external/_main~toolchains~zig_sdk/lib/libcxx/src/barrier.cpp:9:
external/_main~toolchains~zig_sdk/lib/libcxx/include/barrier:51:4: error: "<barrier> is not supported since libc++ has been configured without support for threads."
   51 | #  error "<barrier> is not supported since libc++ has been configured without support for threads."
      |    ^
external/_main~toolchains~zig_sdk/lib/libcxx/src/condition_variable_destructor.cpp:26:3: error: unknown type name '__libcpp_condvar_t'
   26 |   __libcpp_condvar_t __cv_ = _LIBCPP_CONDVAR_INITIALIZER;
      |   ^
1 error generated.
Target //test/c:test_libc_wasip1_wasm32 failed to build
INFO: Elapsed time: 11.475s, Critical Path: 7.98s
INFO: 2 processes: 2 internal.
```

chrisirhc committed:
even newer nightly

chrisirhc committed:
changed more shas

ah-quant commented:
I just want to drop by to say thank you, @chrisirhc! I adapted your existing PR and got a lot further in using Zig as a C/C++ toolchain on Windows (and only Windows - where getting a C/C++ environment set up is just ...blergh).

Please find my patch below - using it, I could continue until I hit https://github.com/ziglang/zig/issues/23233 :-)

I needed 0.14.0 because pthreads.h was added to msys2 after 0.13.0 and I could not get through abseil without it.

If you want to use it, please do - I disclaim any copyright, this is just a minor change to what you did.

```
===================================================================
diff --git a/toolchain/private/zig_sdk.bzl b/toolchain/private/zig_sdk.bzl
index a328df5..c1b3cf5 100644
--- a/toolchain/private/zig_sdk.bzl
+++ b/toolchain/private/zig_sdk.bzl
@@ -1,12 +1,12 @@
-VERSION = "0.12.0"
+VERSION = "0.14.0"

 HOST_PLATFORM_SHA256 = {
+    "windows-x86_64": "f53e5f9011ba20bbc3e0e6d0a9441b31eb227a97bac0e7d24172f1b8b27b4371",
 }

 # Official recommended version. Should use this when we have a usable release.
diff --git a/toolchain/zig-wrapper.zig b/toolchain/zig-wrapper.zig
index d1d59f9..99be373 100644
--- a/toolchain/zig-wrapper.zig
+++ b/toolchain/zig-wrapper.zig
@@ -93,7 +93,7 @@ const ParseResults = union(Action) {
 };

 // sub-commands in the same folder as `zig-wrapper`
-const sub_commands_target = std.ComptimeStringMap(void, .{
+const sub_commands_target = std.StaticStringMap(void).initComptime(.{
     .{"ar"},
     .{"ld.lld"},
     .{"lld-link"},


diff --git a/toolchain/zig-wrapper.zig b/toolchain/zig-wrapper.zig
index 99be373..70ee41e 100644
--- a/toolchain/zig-wrapper.zig
+++ b/toolchain/zig-wrapper.zig
@@ -55,7 +55,7 @@ const std = @import("std");
 const fs = std.fs;
 const mem = std.mem;
 const process = std.process;
-const ChildProcess = std.ChildProcess;
+const ChildProcess = std.process.Child;
 const ArrayListUnmanaged = std.ArrayListUnmanaged;
 const sep = fs.path.sep_str;


diff --git a/toolchain/zig-wrapper.zig b/toolchain/zig-wrapper.zig
index 99be373..70ee41e 100644
--- a/toolchain/zig-wrapper.zig
+++ b/toolchain/zig-wrapper.zig
@@ -279,7 +279,7 @@ const std = @import("std");
     // strings `is.it.x86_64?-stallinux,macos-`; we are trying to aid users
     // that run things from the wrong directory, not trying to punish the ones
     // having fun.
-    var it = mem.split(u8, triple, "-");
+    var it = mem.splitSequence(u8, triple, "-");

     const arch = it.next() orelse return error.BadParent;
     if (mem.indexOf(u8, "aarch64,x86_64,wasm32", arch) == null)

```

chrisirhc committed:
update to 0.14.0

chrisirhc committed:
attempt to resolve mac test issue

chrisirhc committed:
fix error with posix functions

>  The error is coming from Go's CGo trying to build the standard library with the Zig C++ compiler, and it's failing to find certain POSIX terminal functions (grantpt, posix_openpt, ptsname, unlockpt).

```
bazel-out/darwin_x86_64-fastbuild-ST-1dd79e0b6a44/bin/external/rules_go~0.45.1/stdlib_/src/internal/testpty/pty_cgo.go:24:17: could not determine kind of name for C.grantpt
bazel-out/darwin_x86_64-fastbuild-ST-1dd79e0b6a44/bin/external/rules_go~0.45.1/stdlib_/src/internal/testpty/pty_cgo.go:20:12: could not determine kind of name for C.posix_openpt
bazel-out/darwin_x86_64-fastbuild-ST-1dd79e0b6a44/bin/external/rules_go~0.45.1/stdlib_/src/internal/testpty/pty_cgo.go:32:26: could not determine kind of name for C.ptsname
bazel-out/darwin_x86_64-fastbuild-ST-1dd79e0b6a44/bin/external/rules_go~0.45.1/stdlib_/src/internal/testpty/pty_cgo.go:28:17: could not determine kind of name for C.unlockpt
```

chrisirhc committed:
need remaining copts in previous commit

-D_GNU_SOURCE - Enables all GNU extensions
-D_POSIX_C_SOURCE=200809L - Enables POSIX.1-2008 features
-D_XOPEN_SOURCE=600 - Enables X/Open System Interface (XSI) features

These macros should make the necessary POSIX terminal functions (grantpt, posix_openpt, ptsname, unlockpt) available to CGo when building the Go standard library.
The error occurred because these functions are part of the POSIX terminal interface, and they need to be explicitly enabled through feature test macros when building with certain C library implementations.

chrisirhc commented on toolchain/private/defs.bzl:
```diff
@@ -207,12 +211,18 @@ def _target_wasm():
         includes = [
             "libc/include/wasm-wasi-musl",
+            "libc/include/generic-musl",
             "libc/wasi",
         ] + _INCLUDE_TAIL,
         linkopts = [],
         dynamic_library_linkopts = [],
         supports_dynamic_linker = False,
-        copts = [],
+        copts = [
+            "-D_LIBCPP_HAS_MUSL_LIBC",
+            "-D_LIBCPP_HAS_NO_THREADS",
+            "-D_LIBCPP_HAS_NO_THREAD_UNSAFE_C_FUNCTIONS",
+            "-D_LIBCPP_HAS_NO_THREAD_API_PTHREAD",
+        ],
```
Attempt to resolve issue:
```
external/_main~toolchains~zig_config/lib/libc/include/wasm-wasi-musl/stdio.h:8:10: fatal error: 'features.h' file not found
```

chrisirhc replied:
Removing all the additional copts led to the tests passing, so I removed them.

chrisirhc commented on toolchain/private/defs.bzl:
```diff
@@ -161,7 +161,11 @@ def _target_linux_gnu(gocpu, zigcpu, glibc_version):
         dynamic_library_linkopts = [],
         supports_dynamic_linker = True,
-        copts = [],
+        copts = [
+            "-D_GNU_SOURCE",
+            "-D_POSIX_C_SOURCE=200809L",
+            "-D_XOPEN_SOURCE=600",
+        ],
```
Attempt to resolve issue:
```
bazel-out/darwin_x86_64-fastbuild-ST-1dd79e0b6a44/bin/external/rules_go~0.45.1/stdlib_/src/internal/testpty/pty_cgo.go:24:17: could not determine kind of name for C.grantpt
bazel-out/darwin_x86_64-fastbuild-ST-1dd79e0b6a44/bin/external/rules_go~0.45.1/stdlib_/src/internal/testpty/pty_cgo.go:20:12: could not determine kind of name for C.posix_openpt
bazel-out/darwin_x86_64-fastbuild-ST-1dd79e0b6a44/bin/external/rules_go~0.45.1/stdlib_/src/internal/testpty/pty_cgo.go:32:26: could not determine kind of name for C.ptsname
bazel-out/darwin_x86_64-fastbuild-ST-1dd79e0b6a44/bin/external/rules_go~0.45.1/stdlib_/src/internal/testpty/pty_cgo.go:28:17: could not determine kind of name for C.unlockpt
```

chrisirhc replied:
Remove all except for `-D_GNU_SOURCE` led to the tests passing, so I removed them.

chrisirhc commented:
Failure is due to missing ziglang mirror, see https://github.com/bazelbuild/bazel/issues/25575 . Will retry after that has landed.

chrisirhc committed:
fix lint

chrisirhc committed:
reduce to only essential copt

chrisirhc committed:
reduce copts to minimal set

chrisirhc commented on toolchain/zig-wrapper.zig:
```diff
@@ -142,9 +142,9 @@ fn spawnWindows(arena: mem.Allocator, params: ExecParams) u8 {
-        "error spawning {s}: {s}\n",
-        .{ params.args.items[0], @errorName(err) },
-    );
+            "error spawning {s}: {s}\n",
+            .{ params.args.items[0], @errorName(err) },
+        );
```
From new version of zig's zig fmt.

chrisirhc commented:
Do not land. Blocked on https://github.com/ziglang/zig/issues/23287

linzhp commented:
Superseded by #256
